### PR TITLE
Add functions to set firewall rule

### PR DESCRIPTION
**Pull request by cliping**
1. Add functions to setup/cleanup firewall rule.
2. Add a parameter for domjobabort().
3. Update kill_service() to kill_or_control_service().
4. Add need_sleep_time parameter.

Signed-off-by: cliping <lcheng@redhat.com>

### CODE REVIEW

**Comment by cliping** [comment]
Test result in https://github.com/autotest/tp-libvirt/pull/4619.

File: virttest/utils_libvirt/libvirt_network.py [hunk 1]
```diff
@@ -264,3 +265,74 @@ def check_network_connection(net_name, expected_conn=0):
...
+    more quickly after network issue occurs
+
+    :param expected_list: The expected tcp config list
+    """
+    cmd = ("echo %s > /proc/sys/net/ipv4/tcp_keepalive_probes; "
+           "echo %s > /proc/sys/net/ipv4/tcp_keepalive_intvl; "
+           "echo %s > /proc/sys/net/ipv4/tcp_retries1; "
+           "echo %s > /proc/sys/net/ipv4/tcp_retries2; "
+           "echo %s > /proc/sys/net/ipv4/tcp_fin_timeout" %
+           (expected_list["tcp_keepalive_probes"], expected_list["tcp_keepalive_intvl"],
```
Comment (Yingshun) [inline virttest/utils_libvirt/libvirt_network.py:334]: So, is expected_list a dict? Can we have a proper name? And do we need to set default values for these settings?

File: virttest/utils_libvirt/libvirt_service.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+    :param params: Dictionary with the test parameters
+    """
+    service_name = params.get("service_name", "libvirtd")
+    service_on_dst = "yes" == params.get("service_on_dst", "no")
+    server_ip = params.get("remote_ip")
+    server_user = params.get("remote_user", "root")
+    server_pwd = params.get("remote_pwd")
+
+    if service_name == "libvirtd":
```
Comment (Yingshun) [inline virttest/utils_libvirt/libvirt_service.py:23]: Is it better to have a 'else'?

File: virttest/utils_libvirt/libvirt_service.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    """
+    service_on_dst = "yes" == params.get("service_on_dst", "no")
+    service_operations = params.get("service_operations", "restart")
+
+    service_name = get_service_name(params)
+    cmd = "systemctl %s %s" % (service_operations, service_name)
+    if service_on_dst:
+        remote.run_remote_cmd(cmd, params, ignore_status=False)
+    else:
+        process.run(cmd, ignore_status=False, shell=True)
```
Comment (Yingshun) [inline virttest/utils_libvirt/libvirt_service.py:66]: For 'libvirtd', there is a class Libvirtd() to manage libvirtd related service, and for other system services. I would suggest using service.Factory.create_service(service_name, run=runner).

File: virttest/migration.py [hunk 4]
```diff
@@ -453,16 +453,24 @@ def _run_funcs(vm, funcs_to_run, before_pause, virsh_event_session=None):
             :param virsh_event_session: VirshSession to collect domain events
             :raises: exceptions.TestError if any test error happens
             """
+
             for one_func in funcs_to_run:
+                need_sleep_time = one_func.get('need_sleep_time')
                 if isinstance(one_func, (types.FunctionType, types.MethodType)):
                     if not before_pause:
+                        if need_sleep_time:
+                            LOG.debug("Sleep %s." % need_sleep_time)
```
Comment (dzhengfy) [inline virttest/migration.py:462]: `("Sleep %s." , need_sleep_time)`

File: virttest/migration.py [hunk 5]
```diff
@@ -453,16 +453,24 @@ def _run_funcs(vm, funcs_to_run, before_pause, virsh_event_session=None):
...
+                            time.sleep(int(need_sleep_time))
                         _run_simple_func(vm, one_func)
                     else:
                         LOG.error("Only support to run the function "
                                   "after guest is paused")
                 elif isinstance(one_func, dict):
                     before_vm_pause = 'yes' == one_func.get('before_pause', 'no')
                     if before_vm_pause == before_pause:
+                        if need_sleep_time:
+                            LOG.debug("Sleep %s." % need_sleep_time)
```
Comment (dzhengfy) [inline virttest/migration.py:472]: Same as above

File: virttest/utils_libvirt/libvirt_monitor.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+from virttest import virsh
+
+LOG = logging.getLogger('avocado.' + __name__)
+
+
+def check_domjobinfo(params, expected_list, remote=False):
+    """
+    Check domjobinfo
+
+    :param params: dict, get vm name
```
[the remark this answers is not in the document]
Comment (dzhengfy) [reply virttest/utils_libvirt/libvirt_monitor.py:15]: The docstring could be improved as not only vm name is get from params in your codes.

File: virttest/utils_libvirt/libvirt_monitor.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    :param expected_list: expected list
```
Comment (dzhengfy) [inline virttest/utils_libvirt/libvirt_monitor.py:16]: Could you make the docstring more informative?

File: virttest/utils_libvirt/libvirt_network.py [hunk 8]
```diff
@@ -264,3 +265,74 @@ def check_network_connection(net_name, expected_conn=0):
...
+        server_pwd = params.get("server_pwd")
+        remote_session = remote.wait_for_login('ssh', server_ip, '22',
+                                               server_user, server_pwd,
+                                               r"[\#\$]\s*$")
+        firewall_cmd_dest = utils_iptables.Firewall_cmd(remote_session)
+        firewall_cmd_dest.add_direct_rule(firewall_rule_on_dest)
+        remote_session.close()
+
+    if firewall_rule_on_src:
+        LOG.debug("set firewall rule on source: %s" % firewall_rule_on_src)
```
Comment (dzhengfy) [inline virttest/utils_libvirt/libvirt_network.py:292]: %  -> ,

File: virttest/utils_libvirt/libvirt_network.py [hunk 9]
```diff
@@ -264,3 +265,74 @@ def check_network_connection(net_name, expected_conn=0):
...
+    """
+    Setup firewall rule on source/target host
+
+    :param params: dict, get firewall rule on dest and source
+    """
+    firewall_rule_on_dest = params.get("firewall_rule_on_dest")
+    firewall_rule_on_src = params.get("firewall_rule_on_src")
+
+    if firewall_rule_on_dest:
+        LOG.debug("set firewall rule on dest: %s" % firewall_rule_on_dest)
```
Comment (dzhengfy) [inline virttest/utils_libvirt/libvirt_network.py:280]: %  -> ,

File: virttest/utils_libvirt/libvirt_network.py (hunk 1 above)
Comment (dzhengfy) [reply virttest/utils_libvirt/libvirt_network.py:334]: `expected_list["tcp_keepalive_probes"]` might be `expected_list.get("tcp_keepalive_probes", 'xxx') `for better?

File: virttest/utils_libvirt/libvirt_network.py [hunk 10]
```diff
@@ -264,3 +265,74 @@ def check_network_connection(net_name, expected_conn=0):
...
+    """
+    cmd = ("echo %s > /proc/sys/net/ipv4/tcp_keepalive_probes; "
+           "echo %s > /proc/sys/net/ipv4/tcp_keepalive_intvl; "
+           "echo %s > /proc/sys/net/ipv4/tcp_retries1; "
+           "echo %s > /proc/sys/net/ipv4/tcp_retries2; "
+           "echo %s > /proc/sys/net/ipv4/tcp_fin_timeout" %
+           (expected_list["tcp_keepalive_probes"], expected_list["tcp_keepalive_intvl"],
+            expected_list["tcp_retries1"], expected_list["tcp_retries2"],
+            expected_list["tcp_fin_timeout"]))
+    LOG.debug("Change TCP config: %s" % cmd)
```
Comment (dzhengfy) [inline virttest/utils_libvirt/libvirt_network.py:337]: %  -> ,

File: virttest/utils_libvirt/libvirt_service.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    if service_name == "libvirtd":
+        if service_on_dst:
+            remote_session = remote.wait_for_login('ssh', server_ip, '22',
+                                                   server_user, server_pwd,
+                                                   r"[\#\$]\s*$")
+            service_name = utils_libvirtd.Libvirtd(session=remote_session).service_name
+            remote_session.close()
+        else:
+            service_name = utils_libvirtd.Libvirtd().service_name
+    LOG.debug("service name: %s" % service_name)
```
Comment (dzhengfy) [inline virttest/utils_libvirt/libvirt_service.py:32]: % -> ,

File: virttest/utils_libvirt/libvirt_monitor.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (cliping) [reply virttest/utils_libvirt/libvirt_monitor.py]: Updated, thanks.

File: virttest/utils_libvirt/libvirt_service.py (hunk 2 above)
Comment (cliping) [reply virttest/utils_libvirt/libvirt_service.py:26]: Updated.

File: virttest/migration.py (hunk 4 above)
Comment (cliping) [reply virttest/migration.py]: Updated, thanks.

File: virttest/utils_libvirt/libvirt_monitor.py (hunk 7 above)
Comment (cliping) [reply virttest/utils_libvirt/libvirt_monitor.py]: Updated, thanks.

File: virttest/utils_libvirt/libvirt_network.py (hunk 8 above)
Comment (cliping) [reply virttest/utils_libvirt/libvirt_network.py]: Updated.

File: virttest/utils_libvirt/libvirt_network.py (hunk 10 above)
Comment (cliping) [reply virttest/utils_libvirt/libvirt_network.py]: Updated, thanks.

File: virttest/utils_libvirt/libvirt_service.py (hunk 11 above)
Comment (cliping) [reply virttest/utils_libvirt/libvirt_service.py]: Updated.

**Review by Yingshun — APPROVED** [review]

**Review by dzhengfy — APPROVED** [review]

**Pull request merged**